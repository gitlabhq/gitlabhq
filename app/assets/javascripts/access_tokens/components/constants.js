import { __, s__ } from '~/locale';

export const EVENT_ERROR = 'ajax:error';
export const EVENT_SUCCESS = 'ajax:success';
export const FORM_SELECTOR = '#js-new-access-token-form';

export const INITIAL_PAGE = 1;
export const PAGE_SIZE = 100;

const BASE_FIELDS = [
  {
    key: 'name',
    label: __('Token name'),
    isRowHeader: true,
    sortable: true,
  },
  {
    formatter(description) {
      return description ?? '-';
    },
    key: 'description',
    label: __('Description'),
    sortable: true,
  },
  {
    formatter(scopes) {
      return scopes?.length ? scopes.join(', ') : __('no scopes selected');
    },
    key: 'scopes',
    label: __('Scopes'),
    sortable: true,
  },
  {
    key: 'createdAt',
    label: s__('AccessTokens|Created'),
    sortable: true,
  },
  {
    key: 'lastUsedAt',
    label: __('Last Used'),
    sortable: true,
  },
];

const ROLE_FIELD = {
  key: 'role',
  label: __('Role'),
  sortable: true,
};

export const FIELDS = [
  ...BASE_FIELDS,
  {
    formatter(ips) {
      return ips?.length ? ips?.join(', ') : '-';
    },
    key: 'lastUsedIps',
    label: __('Last Used IPs'),
    sortable: false,
    tdClass: 'gl-break-all gl-min-w-20',
  },
  {
    key: 'expiresAt',
    label: __('Expires'),
    sortable: true,
  },
  ROLE_FIELD,
  {
    key: 'action',
    label: __('Action'),
    tdClass: '!gl-py-3',
  },
];

export const INACTIVE_TOKENS_TABLE_FIELDS = [
  ...BASE_FIELDS,
  {
    key: 'expiresAt',
    label: __('Expired'),
    sortable: true,
  },
  ROLE_FIELD,
];
