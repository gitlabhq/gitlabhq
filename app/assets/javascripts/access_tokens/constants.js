import { GlFilteredSearchToken } from '@gitlab/ui';
import { s__, __ } from '~/locale';
import {
  OPERATORS_AFTER_BEFORE,
  OPERATORS_IS,
} from '~/vue_shared/components/filtered_search_bar/constants';
import DateToken from '~/vue_shared/components/filtered_search_bar/tokens/date_token.vue';

// Token types
export const FEED_TOKEN = 'feedToken';
export const INCOMING_EMAIL_TOKEN = 'incomingEmailToken';
export const STATIC_OBJECT_TOKEN = 'staticObjectToken';

export const DEFAULT_FILTER = [
  {
    type: 'state',
    value: {
      data: 'active',
      operator: '=',
    },
  },
];

export const FILTER_OPTIONS = [
  {
    icon: 'status',
    title: s__('AccessTokens|State'),
    type: 'state',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'active', title: s__('AccessTokens|Active') },
      { value: 'inactive', title: s__('AccessTokens|Inactive') },
    ],
  },
  {
    icon: 'remove',
    title: s__('AccessTokens|Revoked'),
    type: 'revoked',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'true', title: __('Yes') },
      { value: 'false', title: __('No') },
    ],
  },
  {
    icon: 'history',
    title: __('Created date'),
    type: 'created',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Expiration date'),
    type: 'expires',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Last used date'),
    type: 'last_used',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
];

export const FILTER_OPTIONS_CREDENTIALS_INVENTORY = [
  {
    icon: 'key',
    title: s__('CredentialsInventory|Type'),
    type: 'filter',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      {
        value: 'personal_access_tokens',
        title: s__('CredentialsInventory|Personal access tokens'),
      },
      { value: 'ssh_keys', title: s__('CredentialsInventory|SSH keys') },
      {
        value: 'resource_access_tokens',
        title: s__('CredentialsInventory|Project and group access tokens'),
      },
      { value: 'gpg_keys', title: s__('CredentialsInventory|GPG keys') },
    ],
  },
  {
    icon: 'user',
    title: s__('CredentialsInventory|Owner type'),
    type: 'owner_type',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'human', title: __('Human') },
      { value: 'service_account', title: __('Service account') },
    ],
  },
  ...FILTER_OPTIONS,
];

export const DEFAULT_SORT = { value: 'expires', isAsc: true };

export const SORT_OPTIONS = [
  {
    text: __('Created date'),
    value: 'created',
    sort: {
      asc: 'created_asc',
      desc: 'created_desc',
    },
  },
  {
    text: __('Expiration date'),
    value: 'expires',
    sort: {
      asc: 'expires_asc',
      desc: 'expires_desc',
    },
  },
  {
    text: __('Last used date'),
    value: 'last_used',
    sort: {
      asc: 'last_used_asc',
      desc: 'last_used_desc',
    },
  },
  {
    text: __('Name'),
    value: 'name',
    sort: {
      asc: 'name_asc',
      desc: 'name_desc',
    },
  },
];

export const STATISTICS_CONFIG = [
  {
    title: s__('AccessTokens|Active tokens'),
    tooltipTitle: s__('AccessTokens|Filter for active tokens'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Tokens expiring in 2 weeks'),
    tooltipTitle: s__('AccessTokens|Filter for tokens expiring in 2 weeks'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
      {
        type: 'expires',
        value: {
          data: 'DATE_HOLDER',
          operator: '<',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Revoked tokens'),
    tooltipTitle: s__('AccessTokens|Filter for revoked tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'true',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Expired tokens'),
    tooltipTitle: s__('AccessTokens|Filter for expired tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'false',
          operator: '=',
        },
      },
      {
        type: 'state',
        value: {
          data: 'inactive',
          operator: '=',
        },
      },
    ],
  },
];
