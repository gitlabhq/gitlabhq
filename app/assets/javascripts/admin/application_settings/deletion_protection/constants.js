import { __, s__ } from '~/locale';

export const I18N_DELETION_PROTECTION = {
  label: s__('DeletionSettings|Retention period'),
  helpText: s__(
    'DeletionSettings|Period that deleted groups and projects will remain restorable for.',
  ),
  learnMore: __('Learn more.'),
  days: __('days'),
};

export const DEL_ADJ_PERIOD_MAX_LIMIT = 90;
export const DEL_ADJ_PERIOD_MIN_LIMIT = 1;

export const DEL_ADJ_PERIOD_MAX_LIMIT_ERROR = s__(
  'DeletionSettings|Maximum deletion protection duration is 90 days.',
);

export const DEL_ADJ_PERIOD_MIN_LIMIT_ERROR = s__(
  'DeletionSettings|Minimum deletion protection duration is 1 day.',
);
