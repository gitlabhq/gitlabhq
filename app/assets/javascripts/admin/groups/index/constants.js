import groupsEmptyStateIllustration from '@gitlab/svgs/dist/illustrations/empty-state/empty-groups-md.svg?url';
import { s__, __ } from '~/locale';
import {
  SORT_LABEL_NAME,
  SORT_LABEL_CREATED,
  SORT_LABEL_UPDATED,
  SORT_LABEL_STORAGE_SIZE,
} from '~/groups_projects/constants';
import GroupsList from '~/vue_shared/components/groups_list/groups_list.vue';
import { formatGraphQLGroups } from '~/vue_shared/components/groups_list/formatter';
import ResourceListsEmptyState from '~/vue_shared/components/resource_lists/empty_state.vue';
import adminGroupsQuery from './graphql/queries/groups.query.graphql';

const baseTab = {
  formatter: (groups) =>
    formatGraphQLGroups(groups, (group) => {
      const adminPath = `/admin/groups/${group.fullPath}`;

      return {
        avatarLabelLink: adminPath,
        editPath: `${adminPath}/edit`,
      };
    }),
  listComponent: GroupsList,
  listComponentProps: {
    listItemClass: 'gl-px-5',
    showGroupIcon: true,
  },
  emptyStateComponent: ResourceListsEmptyState,
  query: adminGroupsQuery,
  queryPath: 'groups',
};

export const ACTIVE_TAB = {
  ...baseTab,
  text: __('Active'),
  value: 'active',
  variables: { active: true },
  countsQueryPath: 'active',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any active groups yet."),
    description: s__(
      'Organization|A group is a collection of several projects. If you organize your projects under a group, it works like a folder.',
    ),
    'data-testid': 'groups-empty-state',
  },
};

export const INACTIVE_TAB = {
  ...baseTab,
  text: __('Inactive'),
  value: 'inactive',
  variables: { active: false },
  countsQueryPath: 'inactive',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any inactive groups."),
    description: s__('Groups|Groups that are archived or pending deletion will appear here.'),
  },
};

export const SORT_OPTION_NAME = {
  value: 'name',
  text: SORT_LABEL_NAME,
};

export const SORT_OPTION_CREATED = {
  value: 'created_at',
  text: SORT_LABEL_CREATED,
};

export const SORT_OPTION_UPDATED = {
  value: 'updated_at',
  text: SORT_LABEL_UPDATED,
};

export const SORT_OPTION_STORAGE_SIZE = {
  value: 'storage_size',
  text: SORT_LABEL_STORAGE_SIZE,
};

export const SORT_OPTIONS = [
  SORT_OPTION_NAME,
  SORT_OPTION_CREATED,
  SORT_OPTION_UPDATED,
  SORT_OPTION_STORAGE_SIZE,
];

export const ADMIN_GROUPS_TABS = [ACTIVE_TAB, INACTIVE_TAB];

export const BASE_ROUTE = '/admin/groups';

export const ADMIN_GROUPS_ROUTE_NAME = 'admin-groups';

export const FIRST_TAB_ROUTE_NAMES = [ADMIN_GROUPS_ROUTE_NAME];

export const FILTERED_SEARCH_TERM_KEY = 'search';
export const FILTERED_SEARCH_NAMESPACE = 'admin-groups';
