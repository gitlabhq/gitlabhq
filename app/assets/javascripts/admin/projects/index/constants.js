import emptyStateProjectsSvgPath from '@gitlab/svgs/dist/illustrations/empty-state/empty-projects-md.svg?url';
import { __, s__ } from '~/locale';
import ProjectsList from '~/vue_shared/components/projects_list/projects_list.vue';
import ResourceListsEmptyState, {
  TYPES,
} from '~/vue_shared/components/resource_lists/empty_state.vue';
import { formatGraphQLProjects } from '~/vue_shared/components/projects_list/formatter';
import projectsQuery from './graphql/queries/projects.query.graphql';

const baseTab = {
  listComponent: ProjectsList,
  listComponentProps: {
    listItemClass: 'gl-px-5',
    showProjectIcon: true,
  },
  emptyStateComponent: ResourceListsEmptyState,
  emptyStateComponentProps: {
    title: s__('Projects|No projects found.'),
    svgPath: emptyStateProjectsSvgPath,
    searchMinimumLength: 3,
    type: TYPES.filter,
  },
  queryPath: 'projects',
  formatter: (projects) =>
    formatGraphQLProjects(projects, (project) => ({
      editPath: `/admin/projects/${project.fullPath}/edit`,
      avatarLabelLink: `/admin/projects/${project.fullPath}`,
    })),
};

export const ACTIVE_TAB = {
  ...baseTab,
  text: __('Active'),
  value: 'active',
  query: projectsQuery,
  variables: { active: true },
  queryPath: 'projects',
  countsQueryPath: 'active',
};

export const INACTIVE_TAB = {
  ...baseTab,
  text: __('Inactive'),
  value: 'inactive',
  query: projectsQuery,
  variables: { active: false },
  queryPath: 'projects',
  countsQueryPath: 'inactive',
  emptyStateComponentProps: {
    ...baseTab.emptyStateComponentProps,
    title: s__("Projects|You don't have any inactive projects."),
    description: s__('Projects|Projects that are archived or pending deletion will appear here.'),
  },
};

export const ADMIN_PROJECTS_TABS = [ACTIVE_TAB, INACTIVE_TAB];

export const BASE_ROUTE = '/admin/projects';

export const ADMIN_PROJECTS_ROUTE_NAME = 'admin-projects';
export const FIRST_TAB_ROUTE_NAMES = [ADMIN_PROJECTS_ROUTE_NAME];

export const FILTERED_SEARCH_NAMESPACE = 'admin';
export const FILTERED_SEARCH_TERM_KEY = 'search';
