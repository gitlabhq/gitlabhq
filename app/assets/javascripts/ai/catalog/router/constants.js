export const AI_CATALOG_INDEX_ROUTE = 'ai-catalog';
