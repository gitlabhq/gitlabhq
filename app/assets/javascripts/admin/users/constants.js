import { GlFilteredSearchToken } from '@gitlab/ui';

import { OPERATORS_IS } from '~/vue_shared/components/filtered_search_bar/constants';
import { s__, __, sprintf } from '~/locale';

export const I18N_USER_ACTIONS = {
  edit: __('Edit'),
  editWithName: (name) => sprintf(__('Edit user: %{user_name}'), { user_name: name }),
  userAdministration: s__('AdminUsers|User administration'),
  unlock: __('Unlock'),
  block: s__('AdminUsers|Block'),
  unblock: s__('AdminUsers|Unblock'),
  approve: s__('AdminUsers|Approve'),
  reject: s__('AdminUsers|Reject'),
  deactivate: s__('AdminUsers|Deactivate'),
  activate: s__('AdminUsers|Activate'),
  ldapBlocked: s__('AdminUsers|Cannot unblock LDAP blocked users'),
  delete: s__('AdminUsers|Delete user'),
  deleteWithContributions: s__('AdminUsers|Delete user and contributions'),
  ban: s__('AdminUsers|Ban user'),
  unban: s__('AdminUsers|Unban user'),
  trust: s__('AdminUsers|Trust user'),
  untrust: s__('AdminUsers|Untrust user'),
};

const createTokenConfig = (options) => ({
  token: GlFilteredSearchToken,
  operators: OPERATORS_IS,
  unique: true,
  ...options,
});

// This is exported so the EE version of constants.js can call it.
export const getTokenConfigs = (accessLevelOptions) => [
  createTokenConfig({
    title: s__('AdminUsers|Access level'),
    type: 'access_level',
    options: accessLevelOptions,
  }),
  createTokenConfig({
    title: __('State'),
    type: 'state',
    options: [
      { value: 'active', title: s__('AdminUsers|Active') },
      { value: 'banned', title: s__('AdminUsers|Banned') },
      { value: 'blocked', title: s__('AdminUsers|Blocked') },
      { value: 'deactivated', title: s__('AdminUsers|Deactivated') },
      { value: 'blocked_pending_approval', title: s__('AdminUsers|Pending approval') },
      { value: 'trusted', title: s__('AdminUsers|Trusted') },
      { value: 'wop', title: s__('AdminUsers|Without projects') },
    ],
  }),
  createTokenConfig({
    title: s__('AdminUsers|Two-factor authentication'),
    type: '2fa',
    options: [
      { value: 'two_factor_enabled', title: __('On') },
      { value: 'two_factor_disabled', title: __('Off') },
    ],
  }),
  createTokenConfig({
    title: __('Type'),
    type: 'type',
    options: [
      { value: 'without_bots', title: s__('AdminUsers|Humans') },
      { value: 'bots', title: s__('AdminUsers|Bots') },
      { value: 'placeholder', title: s__('UserMapping|Placeholder') },
    ],
  }),
  createTokenConfig({
    title: s__('AdminUsers|LDAP sync'),
    type: 'ldap_sync',
    options: [{ value: 'ldap_sync', title: __('True') }],
  }),
];

export const TOKEN_CONFIGS = getTokenConfigs([
  { value: 'admins', title: s__('AdminUsers|Administrator') },
  { value: 'external', title: s__('AdminUsers|External') },
]);

export const SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT = 10;

export const SOLO_OWNED_ORGANIZATIONS_EMPTY = {
  count: 0,
  nodes: [],
};
