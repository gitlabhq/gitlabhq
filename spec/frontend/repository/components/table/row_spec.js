import Vue from 'vue';
import VueApollo from 'vue-apollo';
import { GlBadge, GlLink, GlIcon, GlIntersectionObserver } from '@gitlab/ui';
import { shallowMount, RouterLinkStub } from '@vue/test-utils';
import refQuery from '~/repository/queries/ref.query.graphql';
import { createMockDirective, getBinding } from 'helpers/vue_mock_directive';
import createMockApollo from 'helpers/mock_apollo_helper';
import TableRow from '~/repository/components/table/row.vue';
import FileIcon from '~/vue_shared/components/file_icon.vue';
import { FILE_SYMLINK_MODE } from '~/vue_shared/constants';
import { ROW_APPEAR_DELAY } from '~/repository/constants';

const COMMIT_MOCK = { lockLabel: 'Locked by Root', committedDate: '2019-01-01' };

let wrapper;
let $router;

const createMockApolloProvider = (mockData) => {
  Vue.use(VueApollo);
  const apolloProver = createMockApollo([]);
  apolloProver.clients.defaultClient.cache.writeQuery({ query: refQuery, data: { ...mockData } });

  return apolloProver;
};

function factory({ mockData = { ref: 'main', escapedRef: 'main' }, propsData = {} } = {}) {
  $router = {
    push: jest.fn(),
  };

  wrapper = shallowMount(TableRow, {
    apolloProvider: createMockApolloProvider(mockData),
    propsData: {
      id: '1',
      sha: '0as4k',
      commitInfo: COMMIT_MOCK,
      name: 'name',
      currentPath: 'gitlab-org/gitlab-ce',
      projectPath: 'gitlab-org/gitlab-ce',
      url: `https://test.com`,
      totalEntries: 10,
      rowNumber: 123,
      path: 'gitlab-org/gitlab-ce',
      type: 'tree',
      ...propsData,
    },
    provide: {
      refType: 'heads',
    },
    directives: {
      GlHoverLoad: createMockDirective('gl-hover-load'),
    },
    mocks: {
      $router,
    },
    stubs: {
      RouterLink: RouterLinkStub,
    },
  });
}

describe('Repository table row component', () => {
  const findIcon = () => wrapper.findComponent(GlIcon);
  const findFileIcon = () => wrapper.findComponent(FileIcon);
  const findBadge = () => wrapper.findComponent(GlBadge);
  const findRouterLink = () => wrapper.findComponent(RouterLinkStub);
  const findIntersectionObserver = () => wrapper.findComponent(GlIntersectionObserver);

  it('renders table row', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type: 'file',
        currentPath: '/',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a symlink table row', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type: 'blob',
        currentPath: '/',
        mode: FILE_SYMLINK_MODE,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders table row for path with special character', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test$/test',
        type: 'file',
        currentPath: 'test$',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a gl-hover-load directive', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type: 'blob',
        currentPath: '/',
      },
    });

    const hoverLoadDirective = getBinding(findRouterLink().element, 'gl-hover-load');

    expect(hoverLoadDirective).not.toBeUndefined();
    expect(hoverLoadDirective.value).toBeInstanceOf(Function);
  });

  it.each(['tree', 'blob'])('renders a RouterLink for type $type', (type) => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type,
        currentPath: '/',
      },
    });

    expect(wrapper.findComponent(RouterLinkStub).exists()).toBe(true);
  });

  it('renders a hyperlink for type commit', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type: 'commit',
        currentPath: '/',
      },
    });

    expect(wrapper.find('a').exists()).toBe(true);
  });

  it.each`
    path                   | encodedPath
    ${'test#'}             | ${'test%23'}
    ${'Ã„nderungen'}        | ${'%C3%84nderungen'}
    ${'dir%2f_hello__.sh'} | ${'dir%252f_hello__.sh'}
  `('renders link for $path', ({ path, encodedPath }) => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path,
        type: 'tree',
        currentPath: '/',
      },
    });

    expect(wrapper.findComponent({ ref: 'link' }).props('to')).toBe(
      `/-/tree/main/${encodedPath}?ref_type=heads`,
    );
  });

  it('renders link for directory with hash', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test#',
        type: 'tree',
        currentPath: '/',
      },
    });

    expect(wrapper.find('.tree-item-link').props('to')).toBe(`/-/tree/main/test%23?ref_type=heads`);
  });

  it('renders commit ID for submodule', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type: 'commit',
        currentPath: '/',
      },
    });

    expect(wrapper.find('.commit-sha').text()).toContain('1');
  });

  it('renders link with href', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type: 'blob',
        url: 'https://test.com',
        currentPath: '/',
      },
    });

    expect(wrapper.find('a').attributes('href')).toEqual('https://test.com');
  });

  it('renders LFS badge', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type: 'commit',
        currentPath: '/',
        lfsOid: '1',
      },
    });

    expect(findBadge().exists()).toBe(true);
  });

  it('renders commit and web links with href for submodule', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type: 'commit',
        url: 'https://test.com',
        submoduleTreeUrl: 'https://test.com/commit',
        currentPath: '/',
      },
    });

    expect(wrapper.find('a').attributes('href')).toEqual('https://test.com');
    expect(wrapper.findComponent(GlLink).attributes('href')).toEqual('https://test.com/commit');
  });

  it('renders lock icon', () => {
    factory({
      propsData: {
        id: '1',
        sha: '123',
        path: 'test',
        type: 'tree',
        currentPath: '/',
      },
    });

    expect(findIcon().exists()).toBe(true);
    expect(findIcon().props('name')).toBe('lock');
  });

  it('renders loading icon when path is loading', () => {
    factory({
      propsData: {
        id: '1',
        sha: '1',
        path: 'test',
        type: 'tree',
        currentPath: '/',
        loadingPath: 'test',
      },
    });

    expect(findFileIcon().props('loading')).toBe(true);
  });

  describe('row visibility', () => {
    beforeEach(() => {
      factory({
        propsData: {
          id: '1',
          sha: '1',
          path: 'test',
          type: 'tree',
          currentPath: '/',
          commitInfo: null,
        },
      });
    });

    afterAll(() => jest.useRealTimers());

    it('emits a `row-appear` event', () => {
      const setTimeoutSpy = jest.spyOn(global, 'setTimeout');
      findIntersectionObserver().vm.$emit('appear');

      jest.runAllTimers();

      expect(setTimeoutSpy).toHaveBeenCalledTimes(1);
      expect(setTimeoutSpy).toHaveBeenLastCalledWith(expect.any(Function), ROW_APPEAR_DELAY);
      expect(wrapper.emitted('row-appear')).toEqual([[123]]);
    });
  });
});
