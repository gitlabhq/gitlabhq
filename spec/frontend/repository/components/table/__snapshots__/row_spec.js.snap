// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository table row component renders a symlink table row 1`] = `
<tr
  class="tree-item"
>
  <th
    class="gl-cursor-default gl-font-normal gl-relative tree-item-file-name"
    scope="row"
  >
    <a
      class="gl-inline-flex gl-max-w-full gl-min-w-0 tree-item-link"
      data-testid="file-name-link"
      href="https://test.com"
      title="test"
    >
      <file-icon-stub
        class="gl-mr-2 gl-relative gl-text-secondary"
        cssclasses="gl-relative file-icon"
        filemode="120000"
        filename="test"
        size="16"
      />
      <gl-truncate-stub
        class="gl-grow gl-min-w-0 hover:gl-underline"
        position="end"
        text="test"
        withtooltip="true"
      />
    </a>
    <gl-icon-stub
      arialabel="Locked by Root"
      class="gl-relative gl-z-1"
      name="lock"
      size="12"
      title="Locked by Root"
      variant="current"
    />
  </th>
  <td
    class="cursor-default gl-hidden sm:gl-table-cell tree-commit"
  >
    <gl-link-stub
      class="gl-text-subtle str-truncated-100 tree-commit-link"
    />
    <gl-intersection-observer-stub />
  </td>
  <td
    class="cursor-default gl-text-right gl-text-subtle tree-time-ago"
  >
    <gl-intersection-observer-stub>
      <timeago-tooltip-stub
        cssclass=""
        datetimeformat="asDateTime"
        href=""
        time="2019-01-01"
        tooltipplacement="top"
      />
    </gl-intersection-observer-stub>
  </td>
</tr>
`;

exports[`Repository table row component renders table row 1`] = `
<tr
  class="tree-item"
>
  <th
    class="gl-cursor-default gl-font-normal gl-relative tree-item-file-name"
    scope="row"
  >
    <a
      class="gl-inline-flex gl-max-w-full gl-min-w-0 tree-item-link"
      data-testid="file-name-link"
      href="https://test.com"
      title="test"
    >
      <file-icon-stub
        class="gl-mr-2 gl-relative gl-text-secondary"
        cssclasses="gl-relative file-icon"
        filemode=""
        filename="test"
        size="16"
      />
      <gl-truncate-stub
        class="gl-grow gl-min-w-0 hover:gl-underline"
        position="end"
        text="test"
        withtooltip="true"
      />
    </a>
    <gl-icon-stub
      arialabel="Locked by Root"
      class="gl-relative gl-z-1"
      name="lock"
      size="12"
      title="Locked by Root"
      variant="current"
    />
  </th>
  <td
    class="cursor-default gl-hidden sm:gl-table-cell tree-commit"
  >
    <gl-link-stub
      class="gl-text-subtle str-truncated-100 tree-commit-link"
    />
    <gl-intersection-observer-stub />
  </td>
  <td
    class="cursor-default gl-text-right gl-text-subtle tree-time-ago"
  >
    <gl-intersection-observer-stub>
      <timeago-tooltip-stub
        cssclass=""
        datetimeformat="asDateTime"
        href=""
        time="2019-01-01"
        tooltipplacement="top"
      />
    </gl-intersection-observer-stub>
  </td>
</tr>
`;

exports[`Repository table row component renders table row for path with special character 1`] = `
<tr
  class="tree-item"
>
  <th
    class="gl-cursor-default gl-font-normal gl-relative tree-item-file-name"
    scope="row"
  >
    <a
      class="gl-inline-flex gl-max-w-full gl-min-w-0 tree-item-link"
      data-testid="file-name-link"
      href="https://test.com"
      title="test"
    >
      <file-icon-stub
        class="gl-mr-2 gl-relative gl-text-secondary"
        cssclasses="gl-relative file-icon"
        filemode=""
        filename="test"
        size="16"
      />
      <gl-truncate-stub
        class="gl-grow gl-min-w-0 hover:gl-underline"
        position="end"
        text="test"
        withtooltip="true"
      />
    </a>
    <gl-icon-stub
      arialabel="Locked by Root"
      class="gl-relative gl-z-1"
      name="lock"
      size="12"
      title="Locked by Root"
      variant="current"
    />
  </th>
  <td
    class="cursor-default gl-hidden sm:gl-table-cell tree-commit"
  >
    <gl-link-stub
      class="gl-text-subtle str-truncated-100 tree-commit-link"
    />
    <gl-intersection-observer-stub />
  </td>
  <td
    class="cursor-default gl-text-right gl-text-subtle tree-time-ago"
  >
    <gl-intersection-observer-stub>
      <timeago-tooltip-stub
        cssclass=""
        datetimeformat="asDateTime"
        href=""
        time="2019-01-01"
        tooltipplacement="top"
      />
    </gl-intersection-observer-stub>
  </td>
</tr>
`;
